diff --git a/init/initramfs.c b/init/initramfs.c
index bf3af10c500a..1c74b6d70e5e 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -622,11 +622,11 @@ static int __init populate_rootfs(void)
 {
 	char *err;
 
-	if (do_skip_initramfs) {
+	/*if (do_skip_initramfs) {
 		if (initrd_start)
 			free_initrd();
 		return default_rootfs();
-	}
+	}*/
 
 	err = unpack_to_rootfs(__initramfs_start, __initramfs_size);
 	if (err)
